sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -54,15 +54,18 @@ case class DataSourceV2Relation(
 
   override lazy val metadataOutput: Seq[AttributeReference] = table match {
     case hasMeta: SupportsMetadataColumns =>
-      val resolve = conf.resolver
-      val outputNames = outputSet.map(_.name)
-      def isOutputColumn(col: MetadataColumn): Boolean = {
-        outputNames.exists(name => resolve(col.name, name))
+      getTagValue(CACHED_META_OUTPUT_TAG).getOrElse {
```

Comment:
I'm a bit confused, why do we care about metadata output anymore if it has been added to the output? The column in project will just be resolved as a normal attribute, right?

Reply:
yes, but there is a conflict in `AddMetadataColumns`. We will add a new project(original output),  see
https://github.com/apache/spark/blob/6bb68b5f75ac883423a68583750258656f381c33/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L995-L1002

then with this case, it will be:
```scala
Project (_metadata.file_size)
  Project(c)
    File (_metadata.file_size > 0)
      Relation c, metadataoutput
```

I guess the original idea is we should always use children.metadataoutputs to do resolving to prevent exposing unnecessary metadata columns, isn't it ?

Reply:
Fixed in #39895

Reply:
It's acutally the different issue, my test cases still fail with your pr.

After `addMetadataCol`, the newNode is always different with original node(metadata output merges into output), so an extra project can not be avoid. (Your pr fixs a special case of `hiddenOutputTag` with NaturalJoin)
Then, the only way to resolve rest metadata columns is using metadata output. But before this pr, the metadata output is lost once call `addMetadataCol`.

Reply:
I see, the problem here is we reference metadata col twice in two different nodes. I think the problem here is we add the extra project too early. We should only do it once for the root node.

Reply:
I'm not sure it's safe to add a project for root node. One issue is we do not know the output of root node because it's not resolved.